File: python/src/minaombud/samples/arkivering_sample.py
#
# Sample demonstrating using the API to request packages
# for archiving.
#
import json
import os
import sys
import uuid

import requests

from pathlib import Path
from minaombud import defaults


def sample():
    ### 1. Request API access token
    # The access token should be requested and reused for subsequent requests
    # until it expires at which point a new token must be requested.
    tredjeman = defaults.MINA_OMBUD_TREDJE_MAN
    client_id = defaults.MINA_OMBUD_API_CLIENT_ID
    client_secret = defaults.MINA_OMBUD_API_CLIENT_SECRET
    token_url = defaults.MINA_OMBUD_API_TOKEN_URL
    token_request = {
        "grant_type": "client_credentials",
        "client_id": client_id,
        "client_secret": client_secret,
        "scope": "fullmakt:arkivering",
    }
    token_response = requests.post(token_url, token_request).json()
    access_token = token_response["access_token"]

    ### 2. Invoke API
    # List packages and then download each one.
    api_url = defaults.MINA_OMBUD_API_URL
    paket_url = f"{api_url}/tredjeman/{tredjeman}/arkivering/paket"
    headers = {
        "authorization": f"Bearer {access_token}",
        "x-service-name": "arkivering_sample",
        "x-request-id": str(uuid.uuid4()),
    }
    list_response = requests.get(paket_url, headers=headers)
    content_type = list_response.headers.get("content-type")
    if content_type == "application/json":
        list_response_json = list_response.json()
        print(json.dumps(list_response_json, indent=2))
        output_dir = "archive"
        for pkg in list_response_json["paket"]:
            pkg_id = pkg["id"]
            pkg_name = pkg["namn"]
            zip_dir = os.path.join(output_dir, pkg_name)
            zip_name = f"{pkg_id}.zip"
            Path(zip_dir).mkdir(parents=True, exist_ok=True)
            pkg_response = requests.get(f"{paket_url}/{pkg_id}", headers=headers)
            zip_path = os.path.join(zip_dir, zip_name)
            with open(zip_path, "wb") as f:
                f.write(pkg_response.content)
    else:
        for k, v in list_response.headers.items():
            print(f"{k}: {v}", file=sys.stderr)
        print(file=sys.stderr)
        print(list_response.text, file=sys.stderr)


if __name__ == "__main__":
    sample()

File: dotnet/Samples/ArkiveringSample/ArkiveringSample.cs
using System.Globalization;
using System.Net.Http.Json;
using System.Security.Cryptography;
using System;
using System.IO;
using System.Text;
using System.Text.Json;
using Jose;
using MinaOmbud;
using MinaOmbud.Models;

var defaults = Defaults.Instance;
var authUrl = defaults.MINA_OMBUD_API_TOKEN_URL;
var apiUrl = defaults.MINA_OMBUD_API_URL;
var serializeOptions = JsonCanonicalizer.SerializerOptions;
var acceptableKeyTypes = new HashSet<string> { Jwk.KeyTypes.RSA };
var acceptableCryptos = new HashSet<string> { "RS256", "RS384", "RS512" };
var unauthenticatedClient = new HttpClient();
var authenticatedClient = new HttpClient();
authenticatedClient.Timeout = Timeout.InfiniteTimeSpan;

///////////////////////////////////////////////////////////////////////////////
// 1. Request API access token.
// The access token should be requested and reused for subsequent requests
// until it expires at which point a new token must be requested.

var tokenRequest = new FormUrlEncodedContent(new Dictionary<string, string>
    {
        { "grant_type", "client_credentials" },
        { "client_id", defaults.MINA_OMBUD_API_CLIENT_ID },
        { "client_secret", defaults.MINA_OMBUD_API_CLIENT_SECRET },
        { "scope", "fullmakt:arkivering" }
    }
);

var tokenResponse = await Post<Dictionary<string, object>>(authUrl, tokenRequest);
var accessToken = tokenResponse["access_token"].ToString();
authenticatedClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {accessToken}");
authenticatedClient.DefaultRequestHeaders.Add("x-service-name", "ArkiveringSample.cs");
authenticatedClient.DefaultRequestHeaders.Add("x-request-id", Guid.NewGuid().ToString());

///////////////////////////////////////////////////////////////////////////////
// 2. Invoke API
var Tredjeman = defaults.MINA_OMBUD_TREDJE_MAN;
var PaketUrl = $"{apiUrl}/tredjeman/{Tredjeman}/arkivering/paket";
var response = await ApiGet<ArkiveringsinformationResponse>(PaketUrl);
Console.WriteLine(PrettyPrint(response));

var OutputDir = "archive";

foreach (var pkg in response.Paket)
{
    var id = pkg.Id;
    var ZipName = $"{id}.zip";
    var ZipDir = Path.Join(OutputDir, pkg.Namn);
    var ZipPath = Path.Join(ZipDir, ZipName);
    var ZipData = await GetBytes($"{PaketUrl}/{id}");
    Directory.CreateDirectory(ZipDir);
    File.WriteAllBytes(ZipPath, ZipData);
}

///////////////////////////////////////////////////////////////////////////////
// Utilities
///////////////////////////////////////////////////////////////////////////////
static string PrettyPrint(object value)
{
    return JsonSerializer.Serialize(value, new JsonSerializerOptions(JsonCanonicalizer.SerializerOptions)
    {
        WriteIndented = true
    });
}

async Task<T> Post<T>(string uri, HttpContent value)
{
    var r = await unauthenticatedClient.PostAsync(uri, value);
    r.EnsureSuccessStatusCode();
    var body = await r.Content.ReadFromJsonAsync<T>(serializeOptions);
    if (body == null)
    {
        throw new InvalidDataException($"No response body returned from {uri}");
    }

    return body;
}

async Task<T> ApiGet<T>(string uri)
{
    var r = await authenticatedClient.GetAsync(uri);
    r.EnsureSuccessStatusCode();
    var body = await r.Content.ReadFromJsonAsync<T>(serializeOptions);
    if (body == null)
    {
        throw new InvalidDataException($"No response body returned from {uri}");
    }

    return body;
}

async Task<byte[]> GetBytes(string uri)
{
    var r = await authenticatedClient.GetAsync(uri);
    r.EnsureSuccessStatusCode();
    var body = await r.Content.ReadAsByteArrayAsync();
    if (body == null)
    {
        throw new InvalidDataException($"No response body returned from {uri}");
    }

    return body;
}
